src/main.ts

```diff
@@ -1,127 +1,192 @@
-				return false;
-			}
-			return true;
-		});
-		app.keymap.pushScope(this.navScope);
-
-		console.log('Vim Reading View Navigation loaded.');
-	}
-
-	displayValue(leaf: MarkdownView): string {
-		return leaf.contentEl
-			.getElementsByClassName('markdown-reading-view')[0]
-			.getElementsByClassName('document-search-container')[0]
-			.getCssPropertyValue('display');
-	}
-
-	async onunload() {
-		app.keymap.popScope(this.navScope);
-		console.log('Vim Reading View Navigation unloaded.');
-	}
-
-	getScroll(leaf: MarkdownView): number {
-		return leaf.previewMode.getScroll();
-	}
-
-	scrollDown(leaf: MarkdownView) {
-		const scroll = this.getScroll(leaf);
-		leaf.previewMode.applyScroll(scroll + this.settings.scrollDifference);
-	}
-
-	scrollUp(leaf: MarkdownView) {
-		const scroll = this.getScroll(leaf);
-		leaf.previewMode.applyScroll(scroll - this.settings.scrollDifference);
-	}
-
-	async loadSettings() {
-		this.settings = Object.assign(
-			{},
-			DEFAULT_SETTINGS,
-			await this.loadData()
-		);
-	}
-
-	async saveSettings() {
-		await this.saveData(this.settings);
-	}
+    settings: VimScrollSetting;
+    navScope: Scope;
+
+    async onload() {
+        await this.loadSettings();
+        this.addSettingTab(new VimScrollSettingTab(this.app, this));
+        this.navScope = new Scope(app.scope);
+
+        // Scroll down
+        this.navScope.register([], 'j', (evt: KeyboardEvent) => {
+            const leaf = app.workspace.getActiveViewOfType(MarkdownView);
+            if (
+                leaf.getMode() === 'preview' &&
+                this.displayValue(leaf) === 'none'
+            ) {
+                this.scrollDown(leaf);
+                return false;
+            }
+            return true;
+        });
+
+        // Scroll up
+        this.navScope.register([], 'k', (evt: KeyboardEvent) => {
+            const leaf = app.workspace.getActiveViewOfType(MarkdownView);
+            if (
+                leaf.getMode() === 'preview' &&
+                this.displayValue(leaf) === 'none'
+            ) {
+                this.scrollUp(leaf);
+                return false;
+            }
+            return true;
+        });
+
+        // Jump to top
+        let keyArray: string[] = [];
+        const jumpTopEvent = (event: KeyboardEvent) => {
+            const leaf = app.workspace.getActiveViewOfType(MarkdownView);
+            if (
+                leaf.getMode() === 'preview' &&
+                this.displayValue(leaf) === 'none'
+            ) {
+                event.preventDefault();
+                if (event.key != 'g') {
+                    keyArray = [];
+                } else {
+                    keyArray.push(event.key);
+                    if (keyArray.length === 2) {
+                        leaf.previewMode.applyScroll(0);
+                        keyArray = [];
+                    }
+                }
+            }
+        }
+        addEventListener('keydown', jumpTopEvent, { capture: true });
```

Comment:
Does the event really need to be captured? I think in the original example it made sense because the event listener only got added after the command got invoked, but here it will also get called each time the user types a `g` in the editor or anywhere else, so it shouldn't wait with passing it on until our function gets the active leaf and checks  for preview mode.